move DirectIndexString from Base

bkamins opened the pull request:
PR following https://github.com/JuliaLang/julia/issues/24259

nalimilan commented on src/LegacyStrings.jl:
```diff
@@ -97,4 +98,11 @@ using Compat
         include("rep.jl")
     end
+
+    if isdefined(Base, :DirectIndexString)
```
I think you'll have to define this earlier, since `DirectIndexString` is used above. You'll be able to check that once you'll have a version of Julia without it.

bkamins replied:
moved

ararslan commented:
Looks good to me as well. I assume the UTF16 `BoundsError` is preexisting.

bkamins commented:
Yes - the error seems to be line 763 of `abstractarray.jl` which is called with a number.

ararslan commented:
Are there tests for `DirectIndexString` in Base that could be copied over?

bkamins commented:
I have moved everything from Base that referenced `DirectIndexString`.
No tests used `DirectIndexString` as no type in Base was subtype of `DirectIndexString`.

This part of code:
```
# isvalid(), chr2ind() and ind2chr() for SubString{DirectIndexString}
let ss, s="lorem ipsum",
    sdict=Dict(SubString(s,1,11)=>s,
               SubString(s,1,6)=>"lorem ",
               SubString(s,1,0)=>"",
               SubString(s,2,4)=>"ore",
               SubString(s,2,11)=>"orem ipsum",
               SubString(s,15,14)=>""
               )
    for (ss,s) in sdict
        local ss
        for i in -1:12
            @test isvalid(ss,i)==isvalid(s,i)
        end
    end
    for (ss,s) in sdict
        local ss
        for i in 1:length(ss)
            @test ind2chr(ss,i)==ind2chr(s,i)
        end
    end
    for (ss,s) in sdict
        local ss
        for i in 1:length(ss)
            @test chr2ind(ss,i)==chr2ind(s,i)
        end
    end
end #let
```
seems to have been testing `DirectIndexString` in the past but now it tests `String` containing only ASCII (I have corrected the description in https://github.com/JuliaLang/julia/pull/24262).

ararslan commented:
We should try to get this merged and tagged ASAP to avoid a gap in availability on 0.7. Any ideas about the 0.7 failures, @nalimilan?

nalimilan commented:
Honestly I don't think anybody cares much about `DirectIndexString`... :-)

I've traced the failure to `ReinterpretArray`, see https://github.com/JuliaLang/julia/pull/23750#issuecomment-338503769. The problematic call is `convert(UTF16String, Int16[[0x65, 0x66] [0x67, 0x68]])`, which I really wonder how somebody thought it was worth supporting... The next level of absurdity will be to allow constructing strings from `BitArray{3}`.

nalimilan commented:
I have a fix at https://github.com/JuliaArchive/LegacyStrings.jl/pull/25.

stevengj commented:
`ASCIIString` and `UTF32String` are both subtypes of `DirectIndexString` and get methods like `nextind` and `length` from `DirectIndexString`.  So, the tests for `ASCIIString` and `UTF32String` are implicitly testing `DirectIndexString`.  And people do care about those types.
